Source/FieldSolver/SpectralSolver/SpectralHankelTransformer.cpp

```diff
@@ -0,0 +1,198 @@
+                                                      amrex::FArrayBox const & F_physical,
+                                                      amrex::FArrayBox       & G_spectral)
+{
+    // The Hankel transform is purely real, so the real and imaginary parts of
+    // F can be transformed separately, so a simple loop over components
+    // can be done.
+    // Note that F_physical does not include the imaginary part of mode 0,
+    // but G_spectral does.
+    for (int mode=0 ; mode < m_n_rz_azimuthal_modes ; mode++) {
+        int const mode_r = 2*mode;
+        int const mode_i = 2*mode + 1;
+        if (mode == 0) {
+            int const icomp = 0;
+            dht0[mode]->HankelForwardTransform(F_physical, icomp, G_spectral, mode_r);
+            G_spectral.setVal(0., mode_i);
+        } else {
+            int const icomp = 2*mode - 1;
+            dht0[mode]->HankelForwardTransform(F_physical, icomp  , G_spectral, mode_r);
+            dht0[mode]->HankelForwardTransform(F_physical, icomp+1, G_spectral, mode_i);
+        }
+    }
+}
+
+/* \brief Converts a vector field from the physical to the spectral space for all modes */
+void
+SpectralHankelTransformer::PhysicalToSpectral_Vector (amrex::Box const & box,
+                                                      amrex::FArrayBox & F_r_physical,
+                                                      amrex::FArrayBox & F_t_physical,
+                                                      amrex::FArrayBox & G_p_spectral,
+                                                      amrex::FArrayBox & G_m_spectral)
+{
+    // Note that F and G include the imaginary part of mode 0.
+    // F will be overwritten (by the + and - data).
+
+    using amrex::operator""_rt;
+
+    amrex::Array4<amrex::Real> const & F_r_physical_array = F_r_physical.array();
+    amrex::Array4<amrex::Real> const & F_t_physical_array = F_t_physical.array();
+
+    for (int mode=0 ; mode < m_n_rz_azimuthal_modes ; mode++) {
+
+        int const mode_r = 2*mode;
+        int const mode_i = 2*mode + 1;
+
+        amrex::ParallelFor(box,
+        [=] AMREX_GPU_DEVICE (int i, int j, int k)
+        {
+            amrex::Real const r_real = F_r_physical_array(i,j,k,mode_r);
+            amrex::Real const r_imag = F_r_physical_array(i,j,k,mode_i);
+            amrex::Real const t_real = F_t_physical_array(i,j,k,mode_r);
+            amrex::Real const t_imag = F_t_physical_array(i,j,k,mode_i);
+            // Combine the values
+            // temp_p = (F_r - I*F_t)/2
+            // temp_m = (F_r + I*F_t)/2
+            F_r_physical_array(i,j,k,mode_r) = 0.5_rt*(r_real + t_imag);
+            F_r_physical_array(i,j,k,mode_i) = 0.5_rt*(r_imag - t_real);
+            F_t_physical_array(i,j,k,mode_r) = 0.5_rt*(r_real - t_imag);
+            F_t_physical_array(i,j,k,mode_i) = 0.5_rt*(r_imag + t_real);
+        });
+
+        amrex::Gpu::streamSynchronize();
```

Comment:
I believe your guess is correct, that gemm is running on different streams. Currently, there is no way to control the streams that gemm uses. That is promised for the future, but is not implemented yet. Now, it always creates its own streams.

The stream synchronization points were put in empirically. They are the ones needed in order to get correct results. I know that this isn't an ideal way of handling these types of issues since it maybe system dependent. Perhaps for safety, synchronizations should be done before and after every call to gemm. Note however, that the current code uses the explicit loops when on the GPU and this needs no synchronization points.